sunbeam-python/sunbeam/core/watcher.py

```diff
@@ -0,0 +1,213 @@
+from watcherclient.v1 import client as watcher_client
+
+from sunbeam.core.common import SunbeamException, read_config
+from sunbeam.core.deployment import Deployment
+from sunbeam.core.juju import JujuHelper
+from sunbeam.core.openstack_api import get_admin_connection
+from sunbeam.steps.openstack import REGION_CONFIG_KEY
+
+LOG = logging.getLogger(__name__)
+
+# Timeout of seconds while waiting for the watcher resource to reach the target state.
+WAIT_TIMEOUT = 60 * 3
+# Sleep interval (in seconds) between querying watcher resources.
+WAIT_SLEEP_INTERVAL = 5
+ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME = "Sunbeam Cluster Maintaining Template"
+ENABLE_MAINTENANCE_STRATEGY_NAME = "host_maintenance"
+ENABLE_MAINTENANCE_GOAL_NAME = "cluster_maintaining"
+
+WORKLOAD_BALANCING_GOAL_NAME = "workload_balancing"
+WORKLOAD_BALANCING_STRATEGY_NAME = "workload_stabilization"
+WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME = "Sunbeam Cluster Workload Balancing Template"
+
+
+def get_watcher_client(deployment: Deployment) -> watcher_client.Client:
+    region = read_config(deployment.get_client(), REGION_CONFIG_KEY)["region"]
+    conn = get_admin_connection(
+        jhelper=JujuHelper(deployment.get_connected_controller())
+    )
+
+    watcher_endpoint = conn.session.get_endpoint(
+        service_type="infra-optim",
+        region_name=region,
+    )
+    return watcher_client.Client(session=conn.session, endpoint=watcher_endpoint)
+
+
+def _create_host_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME,
+        description="Audit template for cluster maintaining",
+        goal=ENABLE_MAINTENANCE_GOAL_NAME,
+        strategy=ENABLE_MAINTENANCE_STRATEGY_NAME,
+    )
+    return template
+
+
+def _create_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME,
+        description="Audit template for workload balancing",
+        goal=WORKLOAD_BALANCING_GOAL_NAME,
+        strategy=WORKLOAD_BALANCING_STRATEGY_NAME,
+    )
+    return template
+
+
+def get_enable_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_host_maintenance_audit_template(client=client)
+    return template
+
+
+def get_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_workload_balancing_audit_template(client=client)
+    return template
+
+
+@tenacity.retry(
+    reraise=True,
+    stop=tenacity.stop_after_delay(WAIT_TIMEOUT),
```

Comment:
As discussed, the WAIT_TIMEOUT should adjust based on number of actions it has to perform. Otherwise user might end up to run `sunbeam maintenance` command multiple times if there are more instances to migrate.